### ISSUE TITLE
Ingress Timeout Configurable?

### ISSUE THREAD

**Issue by christophwille** [microsoft/azure-container-apps#597]
### This issue is a: (mark with an x)
- [x] bug report -> please search issues before submitting
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)

We have an ASP.NET 6.0 Web API. The request in question is to a long-running (longer than 4 minutes) endpoint. On ACA we get this:

`{StatusCode: 504, ReasonPhrase: 'Gateway Timeout', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:{  Date: Tue, 24 Jan 2023 09:13:10 GMT  Content-Length: 14  Content-Type: text/plain}}`

Question is: are there ingress timeouts? (or do we need to look elsewhere)

Aside: Are there other limits (eg egress, we also grab quite a bit of data from an external service) documented?

**Comment by DHollenbach** [comment]
I have the same question. https://learn.microsoft.com/en-us/azure/container-apps/ingress?tabs=bash This link suggests the default ingress timeout is set to 240s, and I can't seem to find anything in the docs or via the CLI for configuring this.

**Comment by ahmelsayed** [comment]
The 4 minute idle connection timeout is not configurable.

**Comment by DHollenbach** [comment]
Thank you for confirming. Should we say that we are better off using AKS if a request takes more than 4 minutes? I thought ACA supported long-running tasks e.g., > 4 minutes.

**Comment by peter064226** [comment]
facing the same problem . always return stream timeout error when call an long time running api after 4 minutes , anyone know how to fix it ?

**Comment by merou** [comment]
This was the same with app services in the past - maybe long running jobs should be moved to aca jobs

**Comment by leiftaylorcb** [comment]
Any updates on this?

We are using Azure Container Applications for our organization now.
We've moved about 40 microservices into ACA, and with one of our crucial services we are experiencing this issue.
We may need to move this service back to a virtual machine, or into AKS in the short-term.
**We would greatly value the ingress timeout in the container app environment to be configurable.**

**Comment by ErickPol** [comment]
This obviously represents a very severe constraint which is surprisingly receiving little to no attention from Azure as even the documentation lacks this information - even MS support has trouble understanding it. 
We have just moved 2 large clients from Azure to AWS just because of this inexplicable limitation.
The 240s timeout is completely arbitrary and limiting, so if developing a means of configuring it is indeed difficult I expect Microsoft to expand this maximum arbitrary timeout to at least 1200s as it is not uncommon to see enterprise applications requiring this much time to output a report or analysis.
We already employ container jobs for more complex tasks - in our case timeouts must be greater for user requests where jobs are not really necessary.

**Comment by shivaal-mako** [comment]
+1 any updates on this? This is critical to our application running on Azure Container Apps. We're doing a migration from GCP's Google Cloud Run to Azure Container Apps and desperately need this functionality. Up to 600s would solve the majority of our needs, although we have some applications that occasionally need up to 1800s.

**Comment by HTSchoenfelder** [comment]
I understand the challenges posed by the 240-second timeout for HTTP ingress in Azure Container Apps, which I suspect is the default Envoy timeout. This limitation can hinder operations requiring longer processing times. The timeout is (likely now) mentioned in the official documentation. (https://learn.microsoft.com/en-us/azure/container-apps/ingress-overview#http)

As a proper solution, consider refactoring long-running HTTP requests into asynchronous operations. This approach involves initiating the request and then subscribing to the result, allowing the server to handle the task in the background and notify the client upon completion. This method aligns with best practices for handling extended processes and can mitigate timeout issues. For more details on implementing asynchronous HTTP APIs, you can refer to the Tech Community article: https://techcommunity.microsoft.com/t5/apps-on-azure-blog/asynchronous-http-apis-with-azure-container-apps-jobs/ba-p/4217431

If you cannot wait for the issue's resolution—such as making the timeout configurable—and refactoring isn't feasible, you might explore using TCP ingress as a workaround. This allows HTTP traffic over TCP ingress, bypassing the 240-second timeout. I’ve successfully tested this approach, but note that TCP ingress requires a custom vnet and does not support private endpoints. It’s important to verify that this solution works for your use case and evaluate any potential side effects. (https://learn.microsoft.com/en-us/azure/container-apps/ingress-overview#tcp)

While the TCP workaround might help temporarily, implementing asynchronous request handling is generally the more robust and scalable approach for managing long-running tasks. I hope these insights are helpful.

**Comment by christophwille** [comment]
Given the renaissance of https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events vs WebSockets (one data point only: https://newsletter.pragmaticengineer.com/p/shopify-black-friday) the 240 is becoming quite a nuisance.

**Comment by HTSchoenfelder** [comment]
> Given the renaissance of https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events vs WebSockets (one data point only: https://newsletter.pragmaticengineer.com/p/shopify-black-friday) the 240 is becoming quite a nuisance.

Thank you for bringing up SSE—it’s a very valid point, and I fully agree that the timeout should be made configurable to better support such use cases. It’s especially puzzling why this issue hasn’t received more attention, as it seems like a straightforward problem to address.
That said, I still believe that for most scenarios not relying on SSE, my proposed workaround using TCP ingress could serve as a helpful interim solution and might even work for SSE in some cases.
I also suspect that many developers encounter this issue due to legacy code that relies on long-running requests. In such cases, refactoring to an asynchronous process could still be a better long-term solution, even if it requires additional effort.
I simply wanted to share ideas that could assist others dealing with similar challenges. I hope they’re helpful, even if only as interim solutions.

**Comment by nagendramishr** [comment]
a complex OpenAI request can take a long time,  so this isn't just legacy stuff.

**Comment by yashmochi** [comment]
It's been 2.5 years, any update on this?
My APIs works fine and return response within 50-55 seconds for single request. 
But for scalling, when I run in Container apps concurrent 20-30 requests using JMeter, each request throws 504 after 4 minutes of default timeout.

**Comment by javajenks** [comment]
Bear in mind that remote MCP servers use SSE, so we cannot use ACA to host these containers. This is hardly a legacy technology !! Very frustrating indeed..  

https://techcommunity.microsoft.com/blog/appsonazureblog/host-remote-mcp-servers-in-azure-container-apps/4403550#

**Comment by karpikpl** [comment]
with this being on the roadmap - what does it mean when it comes to timeline?

**Comment by ahmelsayed** [comment]
Should be in preview https://learn.microsoft.com/en-us/azure/container-apps/ingress-environment-configuration?tabs=azure-cli#advanced-ingress-settings

Thanks to @Tratcher

**Comment by Tratcher** [comment]
The portal ux will be available soon. Here's the preview CLI:
```
az containerapp env premium-ingress add -g <group> -n <env> --workload-profile-name <profile> --request-idle-timeout <minutes>
```

**Comment by subhradwip-shipthis** [comment]
> The portal ux will be available soon. Here's the preview CLI:
> 
> ```
> az containerapp env premium-ingress add -g <group> -n <env> --workload-profile-name <profile> --request-idle-timeout <minutes>
> ```

when i am running this getting this issue :  (ManagedEnvironmentIngressConfigurationInvalidWorkloadProfileName) Workload profile name Dedicated is already in use by a container app or job and cannot be used for ingress.

Can you not use the default workload profile?

As we have a Dedicated workload profile D4 for a container app. trying to use that causing that issue.

even tried with the default consumption workload profile still issue.

**Comment by Tratcher** [comment]
@subhradwip-shipthis ingress requires its own workload profile, it won't share with apps. I suggest making a workload profile named ingress-d4.

**Comment by subhradwip-shipthis** [comment]
> [@subhradwip-shipthis](https://github.com/subhradwip-shipthis) ingress requires its own workload profile, it won't share with apps. I suggest making a workload profile named ingress-d4.

@Tratcher I was testing this, and it seems that the timeout is not working properly for the container app. Is it still in development, or do I need to add other configurations? I have set the timeout for a request to 5 or 10 minutes, even though the timeout has been increased from the default one (which is 4 minutes). However, it’s not working as expected. when the request exceeds 5 or 10 minutes, it still keeps going without any timeout error.

**Comment by Tratcher** [comment]
RequestIdleTimeout does not cause timeout errors. This property only applies to requests that have no network activity for the given period (idle), after which the connection is closed without a response. Is your long running request actively sending or receiving data?

**Comment by subhradwip-shipthis** [comment]
> RequestIdleTimeout does not cause timeout errors. This property only applies to requests that have no network activity for the given period (idle), after which the connection is closed without a response. Is your long running request actively sending or receiving data?

Hi @Tratcher thank you for your response.

What I’ve observed is that when I set the `requestIdleTimeout` between **1 minute to 4 minutes**, I receive a **504 Gateway Timeout** error. I tested this by creating a custom endpoint in my API that responds after a specific duration, which I pass as a query parameter. In these cases, when the `requestIdleTimeout` is set to values between 1 and 4 minutes, and I set the endpoint response time to 10 minutes, the request works fine. cause the task not sending any response back or not receiving data from from the long running process, so connection is getting closed. 

However, when I set the `requestIdleTimeout` to **greater than 4 minutes** (e.g., 5 minutes, 6 minutes, or 10 minutes), the request doesn’t time out, even when I configure the endpoint to take longer (e.g., 15 minutes or more). In these situations, I do not receive the 503 Gateway Timeout error, even though the response time set for the endpoint exceeds the request timeout. It should close the connection right? as the long running process is not sending any data.

This is a test endpoint that I’m using to check HTTP timeout behavior, and my goal is to achieve an HTTP request timeout that aligns with the specified timeout duration.

See the response when I set requestIdleTimeout to 2 min and request time set to like 10 min

<img width="558" height="42" alt="Image" src="https://github.com/user-attachments/assets/62698945-6707-4caa-a87e-8d18434e91a1" />

<img width="1924" height="522" alt="Image" src="https://github.com/user-attachments/assets/0027dd07-9f20-45a1-8b8d-2bb568ee2d70" />

this kind of behavior wan to achieve with higher timeout.


If this request idle time is not align with my goal suggest me any other approach
Goal:
want to set a customer time out for Azure Container App.. so that when I send aa request it should return response within that time or throw timeout error.

I upgraded to the  Dedicated workload profile for ingress config. as previously it has default timeout 4 min and some of my request can be longer than the 4 min like 6min or 7min . as a result it is sending me a timeout error.

So want to achieve:
1. increate the default time so the api can have enough time to full fill the request
2. if it is taking more time let say 10 min then fine to have gateway timeout error.. cause no cause no data receieving from that long running process

**Comment by kmadof** [comment]
Is there a way to setup this via bicep/azapi?

**Comment by Tratcher** [comment]
@kmadof like this:
```
resource containerAppsEnvironment 'Microsoft.App/managedEnvironments@2025-02-02-preview' = {
  name: name
  location: location
  tags: tags
  properties: {
	workloadProfiles: [
		{
			name: 'ingresswp'
			workloadProfileType: 'D4'
			minimumCount: 2
			maximumCount: 4
		}
	]
	ingressConfiguration: {
		workloadProfileName: 'ingresswp'
		requestIdleTimeout: 5
	}
  }
}
```

**Comment by subhradwip-shipthis** [comment]
@Tratcher it is working now thank you for your help

**Comment by subhradwip-shipthis** [comment]
@Tratcher 
One doubt if you could help me,

<img width="576" height="186" alt="Image" src="https://github.com/user-attachments/assets/7bb2da7d-c159-4220-b234-517fa28b0d6f" />

can you tell me why the outboundIpAddresses is only one.. For my other container app it has multiple addresses... And is that IP address is being seeing by other external app to which we are making request from the container app.
And if dedicated profile only get one outbound ip address, how can i setup the multiple outboundIpAddress.

**Comment by Tratcher** [comment]
The number of outbound IPs depends on the environment configuration. E.g. if you're using a Consumption Workload Profile you'll see a long list because these pods run on a shared server pool. For a dedicated workload profile you'll only see one IP, the cluster outbound load balancer. 

Yes, other services should see one of these IPs, but be aware the IP can change.

Why do you want multiple outbound addresses? That's something you'd have to contact support about.

**Comment by subhradwip-shipthis** [comment]
> The number of outbound IPs depends on the environment configuration. E.g. if you're using a Consumption Workload Profile you'll see a long list because these pods run on a shared server pool. For a dedicated workload profile you'll only see one IP, the cluster outbound load balancer.
> 
> Yes, other services should see one of these IPs, but be aware the IP can change.
> 
> Why do you want multiple outbound addresses? That's something you'd have to contact support about.

ok for one of my networking project we need the multiple outbound ip.. can we configure them using the public-ip and vnet?.. if not, may we can pull a new service in the same resource group  with consumption workload profile..

**Comment by Tratcher** [comment]
Please contact support for this scenario, it's more involved than we can explore here.

**Comment by GorbadPS3** [comment]
Can this be configured or not...? What's the status

**Comment by Tratcher** [comment]
@GorbadPS3 yes, this has been available in preview for some time and is currently transitioning to GA. 
https://docs.azure.cn/en-us/container-apps/ingress-environment-configuration#advanced-ingress-settings

**Comment by GorbadPS3** [comment]
@Tratcher  Does this require the more expensive Ingress option, rather than the default?

**Comment by Tratcher** [comment]
Yes, since holding onto many idle connections adds resource overhead.
